narwhal/primary/src/tests/certifier_tests.rs

```diff
@@ -230,6 +233,144 @@ async fn propose_header_failure() {
+        .primary_by_id(&id)
+        .unwrap()
+        .to_anemo_address()
+        .unwrap();
+    let network = anemo::Network::bind(own_address)
+        .server_name("narwhal")
+        .private_key(network_key)
+        .start(anemo::Router::new())
+        .unwrap();
+
+    // Set up remote primaries responding with votes.
+    let mut primary_networks = Vec::new();
+    for (i, primary) in fixture.authorities().filter(|a| a.id() != id).enumerate() {
+        let address = committee.primary(&primary.public_key()).unwrap();
+        let name = primary.id();
+        // Create bad signature for a number of byzantines.
+        let vote = if i < num_byzantine {
+            let bad_key: DefinedKeyPair = DefinedKeyPair::generate(&mut StdRng::from_seed([0; 32]));
+            Vote::new_with_signer(&proposed_header, &name, &bad_key)
+        } else {
+            Vote::new_with_signer(&proposed_header, &name, primary.keypair())
+        };
+        let mut mock_server = MockPrimaryToPrimary::new();
+        let mut mock_seq = mockall::Sequence::new();
+        mock_server
+            .expect_request_vote()
+            .times(1)
+            .in_sequence(&mut mock_seq)
+            .return_once(move |_request| {
+                Ok(anemo::Response::new(RequestVoteResponse {
+                    vote: Some(vote),
+                    missing: Vec::new(),
+                }))
+            });
+        let routes = anemo::Router::new().add_rpc_service(PrimaryToPrimaryServer::new(mock_server));
+        primary_networks.push(primary.new_network(routes));
+        println!("New primary added: {:?}", address);
+
+        let address = address.to_anemo_address().unwrap();
+        let peer_id = anemo::PeerId(primary.network_keypair().public().0.to_bytes());
+        network
+            .connect_with_peer_id(address, peer_id)
+            .await
+            .unwrap();
+    }
+
+    // Spawn the core.
+    let synchronizer = Arc::new(Synchronizer::new(
+        id,
+        fixture.committee(),
+        worker_cache.clone(),
+        /* gc_depth */ 50,
+        certificate_store.clone(),
+        payload_store.clone(),
+        tx_certificate_fetcher,
+        tx_new_certificates.clone(),
+        tx_parents.clone(),
+        rx_consensus_round_updates.clone(),
+        rx_synchronizer_network,
+        None,
+        metrics.clone(),
+    ));
+    let _handle = Certifier::spawn(
+        id,
+        committee.clone(),
+        header_store.clone(),
+        certificate_store.clone(),
+        synchronizer,
+        signature_service,
+        tx_shutdown.subscribe(),
+        rx_headers,
+        metrics.clone(),
+        network,
+    );
+
+    // Send a proposed header.
+    let proposed_digest = proposed_header.digest();
+    tx_headers.send(proposed_header).await.unwrap();
+
+    if expect_cert {
+        // A cert is expected, checks that the header digest matches.
+        let certificate = rx_new_certificates.recv().await.unwrap();
```

Comment:
You could also add a timeout here to ensure that test will return early if no certificate gets created within a reasonable amount of time